### ISSUE TITLE
cudaHalf() operator for converting model to cuda FP16

### ISSUE THREAD

**Issue by jorditg** [torch/nn#891]
Module:cuda() converts to model parameters to cuda FP32. 

Should cudaHalf() be implemented? Or is there another way to get the conversion? 

Thanks!

**Comment by nhynes** [comment]
I think that the `cuda` method might need to be reworked. For instance, now that `CudaByteTensor` et al. exist, the data type should be preserved. Right now, there's currently the issue of
```lua
MaskZero, parent = torch.class('nn.MaskZero', 'nn.Module')
function FillZero:__init()
    self.mask = torch.ByteTensor() -- becomes a CudaTensor
end
function FillZero:updateOutput(input)
    self.mask:eq(input, 0) -- still a CudaTensor
    input:maskedFill(self.mask, 0) -- error: expected CudaByteTensor
end

local m = nn.FillZero():cuda()
m:forward(torch.rand(3, 3))
```

**Comment by jorditg** [comment]
Probably this would be one way to do it. An alternative would be to consider CudaHalfTensor and CudaTensor as two GPU computation alternatives for floats (first 16 bit, second 32 bits) and use :cuda() and :cudaHalf() or  :cuda('float' or 'half' or even for generalism 'double') to make the calculations in GPU, understanding that all the module variables different from floating point remain the same and only floating points are converted to one or another.

**Comment by nhynes** [comment]
Yes, that sounds right. The only modification I'd make to that proposal is to *also* ship the non-floating-point tensors to the GPU while also preserving the type (i.e. `LongTensor`->`CudaLongTensor`). Whereas you can already do `cudaHalf` using `module:type('torch.CudaHalfTensor')`, that would just turn all tensors into HalfTensors. Adding a prefix would require changing `nn.utils.recursiveType`.

Perhaps I should make a new issue since we seem to be describing different problems!

**Comment by jorditg** [comment]
Hi again. I'm doing dome tests modifying Module.lua and I found better for the portability of the code to use the same method cuda() instead of a different one (cudaHalf()) introducing a optional parameter for switching between FP32 or FP16 conversion. I think that even better would be to have a setdefaultcudatensor() function. In this way, calling :cuda() function would make the conversion to the default tensor.

**Comment by jorditg** [comment]
Anyway, I make the conversion of the model to CudaHalfTensor using module:type('torch.CudaHalfTensor') and after calling: 

local parameters, gradParameters = model:getParameters()

I've got the next error:

Module.lua:256: bad argument #1 to 'set' (expecting number or Tensor or Storage)
stack traceback:
	[C]: in function 'set'
	/home/jordi/torch/install/share/lua/5.1/nn/Module.lua:256: in function 'flatten'
	/home/jordi/torch/install/share/lua/5.1/nn/Module.lua:296: in function 'getParameters'

Any idea of what the problem is?

Thanks!

**Comment by nhynes** [comment]
Well, for one, you seem to be using an outdated version of `nn`, so it might help to update it.

Since it's not clear to which `set` call the stack trace refers, my best guesses are that
* your model didn't return any parameters and your version of flatten has no [guard](https://github.com/torch/nn/blob/master/Module.lua#L205)
* one of your modules has a tensor with no storage

It would be immensely helpful for debugging if you could update your nn and repost the stack trace!

**Comment by jorditg** [comment]
Let's see if I'm able to explain the changes and the error:

1) First of all, I update torch:
Inside the torch directory I run:
./update.sh
./install.sh

2) Inside Module.lua I modify the function cuda() in this way:
```lua
function Module:cuda(mode, ...)
   local mode = mode or 'float'
   if mode == 'half' then
      return self:type('torch.CudaHalfTensor',...)
   end
   return self:type('torch.CudaTensor',...)
end
```
Inside Criterion.lua:
```lua
function Criterion:cuda(mode)
   local mode = mode or 'float'
   if mode == 'half' then
      return self:type('torch.CudaHalfTensor')
   end
   return self:type('torch.CudaTensor')
end
```
With this modifications,  using mode  != 'half' the code works perfectly well. 

Using mode == 'half' the full stack trace  I get is: 

```
/home/jordi/torch/install/bin/luajit: /home/jordi/torch/install/share/lua/5.1/nn/Module.lua:257: bad argument #1 to 'set' (expecting number or Tensor or Storage)
stack traceback:
	[C]: in function 'set'
	/home/jordi/torch/install/share/lua/5.1/nn/Module.lua:257: in function 'flatten'
	/home/jordi/torch/install/share/lua/5.1/nn/Module.lua:297: in function 'getParameters'
	/fast/retine/trainer/train.lua:127: in main chunk
	[C]: in function 'dofile'
	main.lua:44: in main chunk
	[C]: in function 'dofile'
	...ordi/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00405d50
``` 

train.lua:127 is the line:
```
local parameters, gradParameters = model:getParameters()
```
that works perfectly well with fp32.

Thanks,

Jordi

**Comment by gchanan** [comment]
cudaHalf/cudaDouble now exist if you pull in cunn.
